internal/cmd/io_binary.go

```diff
@@ -0,0 +1,276 @@
+}
+
+type binaryIO struct {
+	cmd 					*exec.Cmd
+
+	stdin, stdout, stderr 	string
+
+	sin 					io.Reader
+	sinCloser				sync.Once
+	binaryCloser	sync.Once
+
+	sout, serr, wait 		io.ReadWriteCloser
+	soutCloser				sync.Once
+
+	terminal 				bool
+}
+
+func NewBinaryIO(ctx context.Context, id string, uri *url.URL) (UpstreamIO, error) {
+	ns, err := namespaces.NamespaceRequired(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		sout, serr, w io.ReadWriteCloser = nil, nil, nil
+		stderrPipe, waitPipe string
+	)
+
+	cmd := NewBinaryCmd(ctx, uri, id, ns)
+
+	stdoutPipe :=fmt.Sprintf(`%s\binary-%s-stdout`, pipeRootBinary, id)
+
+	sout, err = openNPipe(stdoutPipe)
+	if err != nil {
+		return nil, err
+	}
+
+	// stderrPipe := fmt.Sprintf(`%s\binary-%s-stderr`, pipeRootBinary, id)
+	// serr, err = openNPipe(stderrPipe)
+	// if err != nil {
+	// 	return nil, err
+	// }
+
+	waitPipe = fmt.Sprintf(`%s\binary-%s-wait`, pipeRootBinary, id)
+	w, err = openNPipe(waitPipe)
+	if err != nil {
+		return nil, err
+	}
+
+	cmd.Env = append(cmd.Env,
+		"CONTAINER_STDOUT="+stdoutPipe,
+		"CONTAINER_STDERR="+stderrPipe,
+		"CONTAINER_WAIT="+waitPipe,
+	)
+
+	started := make(chan bool, 1)
+	if err := cmd.Start(); err != nil {
+		return nil, err
+	}
+
+	go func() {
+		defer w.Close()
+		t := 0
+		for {
+			b := make([]byte, 1)
+			v, err := w.Read(b)
+			if (v==0) || err !=nil && err != io.EOF {
+				log.G(ctx).Debugf("Failed to read from wait pipe. Sleeping")
+				time.Sleep(1 * time.Second)
+				t++
+				if t > 10 {
+					break
+				}
+			} else {
+				log.G(ctx).Debugf("Read from wait pipe. Binary started: %s", b)
+				started <- true
+				return
+			}
+		}
+		started <- false
+	}()
+
+	if !<-started {
+		return nil, errors.Errorf("Failed to started binary")
+	}
+
+	return &binaryIO{
+		cmd:		cmd,
+		stdout:		stdoutPipe,
+		stderr:		stderrPipe,
+		sin:		nil,
+		sout: 		sout,
+		// TODO replace with working pipe
+		serr:		serr,
+		terminal:	false,
+		wait:		w,
+	}, nil
+}
+
+var _ UpstreamIO= &binaryIO{}
+
+func (b *binaryIO) Close(ctx context.Context) {
+	b.cmd.Wait()
+
+	// b.binaryCloser.Do(func() {
+	// 	// Borrowed this from Ming's PR
+	// 	log.G(ctx).Debugf("Waiting for binaryIO to exit: %d", b.cmd.Process.Pid)
+	// 	done := make(chan error, 1)
+	// 	go func() {
+	// 		done <- b.cmd.Wait()
+	// 	}()
+
+	// 	select {
+	// 	case err := <- done:
+	// 		log.G(ctx).WithError(err).Errorf("Error while waiting for cmd to finish")
+	// 	case <-time.After(5*time.Second):
+	// 		log.G(ctx).Errorf("Timeout while waiting for binaryIO process to finish")
+	// 		err := b.cmd.Process.Kill()
+	// 		if err != nil {
+	// 			log.G(ctx).WithError(err).Errorf("Error while killing binaryIO process")
+	// 		}
+	// 	}
+	// })
+	b.soutCloser.Do(func() {
+		if b.sout != nil {
+			log.G(ctx).Debug("binaryIO::soutCloser - stdout")
+			err := b.sout.Close()
+			if err != nil {
+				log.G(ctx).WithError(err).Errorf("Error while closing stdout npipe")
+			}
+		}
+		if b.serr != nil {
```

Comment:
should we set b.serr to nil after successfully closing?

Reply:
is that what we usually do? or is it to make sure that close doesn't get called more than once?